if not game then script = require "test/relative-string" end
local typeof = game and typeof or require "test/mock".typeof:: never
local Instance = game and Instance or require "test/mock".Instance :: never

local throw = require(script.Parent.throw)
local defaults = require(script.Parent.defaults)
local apply = require(script.Parent.apply)
local r = require(script.Parent.roblox_types)

local ctor_cache = {} :: { [string]: () -> Instance }

setmetatable(ctor_cache :: any, {
    __index = function(self, class)
        local ok, instance: Instance = pcall(Instance.new, class :: any)
        if not ok then throw(`invalid class name, could not create instance of class { class }`) end

        local default: { [string]: unknown }? = defaults[class]
        if default then
            for i, v in next, default do
                (instance :: any)[i] = v
            end
        end

        local function ctor(properties: Props): Instance
            return apply(instance:Clone(), properties)    
        end  

        self[class] = ctor
        return ctor
    end
})

local function create_instance(class: string)
    return ctor_cache[class]
end

local function clone_instance(instance: Instance)
    return function(properties: Props): Instance
        local clone = instance:Clone()
        if not clone then throw "attempt to clone a non-archivable instance" end
        return apply(clone, properties)
    end
end

local function create(class_or_instance: string|Instance): (Props) -> Instance
    if type(class_or_instance) == "string" then
        return create_instance(class_or_instance)
    elseif typeof(class_or_instance) == "Instance" then
        return clone_instance(class_or_instance)
    else
        throw("bad argument #1, expected string or instance, got " .. typeof(class_or_instance))
        return nil :: never
    end
end

type Props = { [any]: any }
return (create :: any) :: 
	( (class: "AccessoryDescription") -> (r.vAccessoryDescription) -> AccessoryDescription )
&	( (class: "Accoutrement") -> (r.vAccoutrement) -> Accoutrement )
&	( (class: "Accessory") -> (r.vAccessory) -> Accessory )
&	( (class: "Hat") -> (r.vHat) -> Hat )
&	( (class: "AdPortal") -> (r.vAdPortal) -> AdPortal )
&	( (class: "AdvancedDragger") -> (r.vAdvancedDragger) -> AdvancedDragger )
&	( (class: "AnalyticsService") -> (r.vAnalyticsService) -> AnalyticsService )
&	( (class: "Animation") -> (r.vAnimation) -> Animation )
&	( (class: "CurveAnimation") -> (r.vCurveAnimation) -> CurveAnimation )
&	( (class: "KeyframeSequence") -> (r.vKeyframeSequence) -> KeyframeSequence )
&	( (class: "AnimationController") -> (r.vAnimationController) -> AnimationController )
&	( (class: "AnimationRigData") -> (r.vAnimationRigData) -> AnimationRigData )
&	( (class: "Animator") -> (r.vAnimator) -> Animator )
&	( (class: "Atmosphere") -> (r.vAtmosphere) -> Atmosphere )
&	( (class: "Attachment") -> (r.vAttachment) -> Attachment )
&	( (class: "Bone") -> (r.vBone) -> Bone )
&	( (class: "AudioAnalyzer") -> (r.vAudioAnalyzer) -> AudioAnalyzer )
&	( (class: "AudioChorus") -> (r.vAudioChorus) -> AudioChorus )
&	( (class: "AudioCompressor") -> (r.vAudioCompressor) -> AudioCompressor )
&	( (class: "AudioDeviceInput") -> (r.vAudioDeviceInput) -> AudioDeviceInput )
&	( (class: "AudioDeviceOutput") -> (r.vAudioDeviceOutput) -> AudioDeviceOutput )
&	( (class: "AudioDistortion") -> (r.vAudioDistortion) -> AudioDistortion )
&	( (class: "AudioEcho") -> (r.vAudioEcho) -> AudioEcho )
&	( (class: "AudioEmitter") -> (r.vAudioEmitter) -> AudioEmitter )
&	( (class: "AudioEqualizer") -> (r.vAudioEqualizer) -> AudioEqualizer )
&	( (class: "AudioFader") -> (r.vAudioFader) -> AudioFader )
&	( (class: "AudioFlanger") -> (r.vAudioFlanger) -> AudioFlanger )
&	( (class: "AudioListener") -> (r.vAudioListener) -> AudioListener )
&	( (class: "AudioPitchShifter") -> (r.vAudioPitchShifter) -> AudioPitchShifter )
&	( (class: "AudioPlayer") -> (r.vAudioPlayer) -> AudioPlayer )
&	( (class: "AudioReverb") -> (r.vAudioReverb) -> AudioReverb )
&	( (class: "AudioSearchParams") -> (r.vAudioSearchParams) -> AudioSearchParams )
&	( (class: "Backpack") -> (r.vBackpack) -> Backpack )
&	( (class: "RemoteEvent") -> (r.vRemoteEvent) -> RemoteEvent )
&	( (class: "UnreliableRemoteEvent") -> (r.vUnreliableRemoteEvent) -> UnreliableRemoteEvent )
&	( (class: "WrapLayer") -> (r.vWrapLayer) -> WrapLayer )
&	( (class: "WrapTarget") -> (r.vWrapTarget) -> WrapTarget )
&	( (class: "Beam") -> (r.vBeam) -> Beam )
&	( (class: "BindableEvent") -> (r.vBindableEvent) -> BindableEvent )
&	( (class: "BindableFunction") -> (r.vBindableFunction) -> BindableFunction )
&	( (class: "BodyAngularVelocity") -> (r.vBodyAngularVelocity) -> BodyAngularVelocity )
&	( (class: "BodyForce") -> (r.vBodyForce) -> BodyForce )
&	( (class: "BodyGyro") -> (r.vBodyGyro) -> BodyGyro )
&	( (class: "BodyPosition") -> (r.vBodyPosition) -> BodyPosition )
&	( (class: "BodyThrust") -> (r.vBodyThrust) -> BodyThrust )
&	( (class: "BodyVelocity") -> (r.vBodyVelocity) -> BodyVelocity )
&	( (class: "RocketPropulsion") -> (r.vRocketPropulsion) -> RocketPropulsion )
&	( (class: "BodyPartDescription") -> (r.vBodyPartDescription) -> BodyPartDescription )
&	( (class: "Breakpoint") -> (r.vBreakpoint) -> Breakpoint )
&	( (class: "BubbleChatMessageProperties") -> (r.vBubbleChatMessageProperties) -> BubbleChatMessageProperties )
&	( (class: "Camera") -> (r.vCamera) -> Camera )
&	( (class: "BodyColors") -> (r.vBodyColors) -> BodyColors )
&	( (class: "CharacterMesh") -> (r.vCharacterMesh) -> CharacterMesh )
&	( (class: "Pants") -> (r.vPants) -> Pants )
&	( (class: "Shirt") -> (r.vShirt) -> Shirt )
&	( (class: "ShirtGraphic") -> (r.vShirtGraphic) -> ShirtGraphic )
&	( (class: "Skin") -> (r.vSkin) -> Skin )
&	( (class: "ClickDetector") -> (r.vClickDetector) -> ClickDetector )
&	( (class: "DragDetector") -> (r.vDragDetector) -> DragDetector )
&	( (class: "Clouds") -> (r.vClouds) -> Clouds )
&	( (class: "Configuration") -> (r.vConfiguration) -> Configuration )
&	( (class: "AlignOrientation") -> (r.vAlignOrientation) -> AlignOrientation )
&	( (class: "AlignPosition") -> (r.vAlignPosition) -> AlignPosition )
&	( (class: "AngularVelocity") -> (r.vAngularVelocity) -> AngularVelocity )
&	( (class: "AnimationConstraint") -> (r.vAnimationConstraint) -> AnimationConstraint )
&	( (class: "BallSocketConstraint") -> (r.vBallSocketConstraint) -> BallSocketConstraint )
&	( (class: "HingeConstraint") -> (r.vHingeConstraint) -> HingeConstraint )
&	( (class: "LineForce") -> (r.vLineForce) -> LineForce )
&	( (class: "LinearVelocity") -> (r.vLinearVelocity) -> LinearVelocity )
&	( (class: "PlaneConstraint") -> (r.vPlaneConstraint) -> PlaneConstraint )
&	( (class: "Plane") -> (r.vPlane) -> Plane )
&	( (class: "RigidConstraint") -> (r.vRigidConstraint) -> RigidConstraint )
&	( (class: "RodConstraint") -> (r.vRodConstraint) -> RodConstraint )
&	( (class: "RopeConstraint") -> (r.vRopeConstraint) -> RopeConstraint )
&	( (class: "CylindricalConstraint") -> (r.vCylindricalConstraint) -> CylindricalConstraint )
&	( (class: "PrismaticConstraint") -> (r.vPrismaticConstraint) -> PrismaticConstraint )
&	( (class: "SpringConstraint") -> (r.vSpringConstraint) -> SpringConstraint )
&	( (class: "Torque") -> (r.vTorque) -> Torque )
&	( (class: "TorsionSpringConstraint") -> (r.vTorsionSpringConstraint) -> TorsionSpringConstraint )
&	( (class: "UniversalConstraint") -> (r.vUniversalConstraint) -> UniversalConstraint )
&	( (class: "VectorForce") -> (r.vVectorForce) -> VectorForce )
&	( (class: "HumanoidController") -> (r.vHumanoidController) -> HumanoidController )
&	( (class: "SkateboardController") -> (r.vSkateboardController) -> SkateboardController )
&	( (class: "VehicleController") -> (r.vVehicleController) -> VehicleController )
&	( (class: "AirController") -> (r.vAirController) -> AirController )
&	( (class: "ClimbController") -> (r.vClimbController) -> ClimbController )
&	( (class: "GroundController") -> (r.vGroundController) -> GroundController )
&	( (class: "SwimController") -> (r.vSwimController) -> SwimController )
&	( (class: "ControllerManager") -> (r.vControllerManager) -> ControllerManager )
&	( (class: "CustomEvent") -> (r.vCustomEvent) -> CustomEvent )
&	( (class: "CustomEventReceiver") -> (r.vCustomEventReceiver) -> CustomEventReceiver )
&	( (class: "BlockMesh") -> (r.vBlockMesh) -> BlockMesh )
&	( (class: "CylinderMesh") -> (r.vCylinderMesh) -> CylinderMesh )
&	( (class: "EditableMesh") -> (r.vEditableMesh) -> EditableMesh )
&	( (class: "FileMesh") -> (r.vFileMesh) -> FileMesh )
&	( (class: "SpecialMesh") -> (r.vSpecialMesh) -> SpecialMesh )
&	( (class: "DataStoreGetOptions") -> (r.vDataStoreGetOptions) -> DataStoreGetOptions )
&	( (class: "DataStoreIncrementOptions") -> (r.vDataStoreIncrementOptions) -> DataStoreIncrementOptions )
&	( (class: "DataStoreOptions") -> (r.vDataStoreOptions) -> DataStoreOptions )
&	( (class: "DataStoreSetOptions") -> (r.vDataStoreSetOptions) -> DataStoreSetOptions )
&	( (class: "DebuggerWatch") -> (r.vDebuggerWatch) -> DebuggerWatch )
&	( (class: "Dialog") -> (r.vDialog) -> Dialog )
&	( (class: "DialogChoice") -> (r.vDialogChoice) -> DialogChoice )
&	( (class: "Dragger") -> (r.vDragger) -> Dragger )
&	( (class: "EditableImage") -> (r.vEditableImage) -> EditableImage )
&	( (class: "RobloxEditableImage") -> (r.vRobloxEditableImage) -> RobloxEditableImage )
&	( (class: "EulerRotationCurve") -> (r.vEulerRotationCurve) -> EulerRotationCurve )
&	( (class: "ExperienceInviteOptions") -> (r.vExperienceInviteOptions) -> ExperienceInviteOptions )
&	( (class: "Explosion") -> (r.vExplosion) -> Explosion )
&	( (class: "FaceControls") -> (r.vFaceControls) -> FaceControls )
&	( (class: "Decal") -> (r.vDecal) -> Decal )
&	( (class: "Texture") -> (r.vTexture) -> Texture )
&	( (class: "Hole") -> (r.vHole) -> Hole )
&	( (class: "MotorFeature") -> (r.vMotorFeature) -> MotorFeature )
&	( (class: "Fire") -> (r.vFire) -> Fire )
&	( (class: "FloatCurve") -> (r.vFloatCurve) -> FloatCurve )
&	( (class: "FlyweightService") -> (r.vFlyweightService) -> FlyweightService )
&	( (class: "CSGDictionaryService") -> (r.vCSGDictionaryService) -> CSGDictionaryService )
&	( (class: "NonReplicatedCSGDictionaryService") -> (r.vNonReplicatedCSGDictionaryService) -> NonReplicatedCSGDictionaryService )
&	( (class: "Folder") -> (r.vFolder) -> Folder )
&	( (class: "ForceField") -> (r.vForceField) -> ForceField )
&	( (class: "FunctionalTest") -> (r.vFunctionalTest) -> FunctionalTest )
&	( (class: "GetTextBoundsParams") -> (r.vGetTextBoundsParams) -> GetTextBoundsParams )
&	( (class: "CanvasGroup") -> (r.vCanvasGroup) -> CanvasGroup )
&	( (class: "Frame") -> (r.vFrame) -> Frame )
&	( (class: "ImageButton") -> (r.vImageButton) -> ImageButton )
&	( (class: "TextButton") -> (r.vTextButton) -> TextButton )
&	( (class: "ImageLabel") -> (r.vImageLabel) -> ImageLabel )
&	( (class: "TextLabel") -> (r.vTextLabel) -> TextLabel )
&	( (class: "ScrollingFrame") -> (r.vScrollingFrame) -> ScrollingFrame )
&	( (class: "TextBox") -> (r.vTextBox) -> TextBox )
&	( (class: "VideoFrame") -> (r.vVideoFrame) -> VideoFrame )
&	( (class: "ViewportFrame") -> (r.vViewportFrame) -> ViewportFrame )
&	( (class: "BillboardGui") -> (r.vBillboardGui) -> BillboardGui )
&	( (class: "ScreenGui") -> (r.vScreenGui) -> ScreenGui )
&	( (class: "GuiMain") -> (r.vGuiMain) -> GuiMain )
&	( (class: "AdGui") -> (r.vAdGui) -> AdGui )
&	( (class: "SurfaceGui") -> (r.vSurfaceGui) -> SurfaceGui )
&	( (class: "FloorWire") -> (r.vFloorWire) -> FloorWire )
&	( (class: "SelectionBox") -> (r.vSelectionBox) -> SelectionBox )
&	( (class: "BoxHandleAdornment") -> (r.vBoxHandleAdornment) -> BoxHandleAdornment )
&	( (class: "ConeHandleAdornment") -> (r.vConeHandleAdornment) -> ConeHandleAdornment )
&	( (class: "CylinderHandleAdornment") -> (r.vCylinderHandleAdornment) -> CylinderHandleAdornment )
&	( (class: "ImageHandleAdornment") -> (r.vImageHandleAdornment) -> ImageHandleAdornment )
&	( (class: "LineHandleAdornment") -> (r.vLineHandleAdornment) -> LineHandleAdornment )
&	( (class: "SphereHandleAdornment") -> (r.vSphereHandleAdornment) -> SphereHandleAdornment )
&	( (class: "WireframeHandleAdornment") -> (r.vWireframeHandleAdornment) -> WireframeHandleAdornment )
&	( (class: "ParabolaAdornment") -> (r.vParabolaAdornment) -> ParabolaAdornment )
&	( (class: "SelectionSphere") -> (r.vSelectionSphere) -> SelectionSphere )
&	( (class: "ArcHandles") -> (r.vArcHandles) -> ArcHandles )
&	( (class: "Handles") -> (r.vHandles) -> Handles )
&	( (class: "SurfaceSelection") -> (r.vSurfaceSelection) -> SurfaceSelection )
&	( (class: "SelectionPartLasso") -> (r.vSelectionPartLasso) -> SelectionPartLasso )
&	( (class: "SelectionPointLasso") -> (r.vSelectionPointLasso) -> SelectionPointLasso )
&	( (class: "Path2D") -> (r.vPath2D) -> Path2D )
&	( (class: "HeightmapImporterService") -> (r.vHeightmapImporterService) -> HeightmapImporterService )
&	( (class: "HiddenSurfaceRemovalAsset") -> (r.vHiddenSurfaceRemovalAsset) -> HiddenSurfaceRemovalAsset )
&	( (class: "Highlight") -> (r.vHighlight) -> Highlight )
&	( (class: "Humanoid") -> (r.vHumanoid) -> Humanoid )
&	( (class: "HumanoidDescription") -> (r.vHumanoidDescription) -> HumanoidDescription )
&	( (class: "IKControl") -> (r.vIKControl) -> IKControl )
&	( (class: "RotateP") -> (r.vRotateP) -> RotateP )
&	( (class: "RotateV") -> (r.vRotateV) -> RotateV )
&	( (class: "Glue") -> (r.vGlue) -> Glue )
&	( (class: "ManualGlue") -> (r.vManualGlue) -> ManualGlue )
&	( (class: "ManualWeld") -> (r.vManualWeld) -> ManualWeld )
&	( (class: "Motor") -> (r.vMotor) -> Motor )
&	( (class: "Motor6D") -> (r.vMotor6D) -> Motor6D )
&	( (class: "Rotate") -> (r.vRotate) -> Rotate )
&	( (class: "Snap") -> (r.vSnap) -> Snap )
&	( (class: "VelocityMotor") -> (r.vVelocityMotor) -> VelocityMotor )
&	( (class: "Weld") -> (r.vWeld) -> Weld )
&	( (class: "Keyframe") -> (r.vKeyframe) -> Keyframe )
&	( (class: "KeyframeMarker") -> (r.vKeyframeMarker) -> KeyframeMarker )
&	( (class: "PointLight") -> (r.vPointLight) -> PointLight )
&	( (class: "SpotLight") -> (r.vSpotLight) -> SpotLight )
&	( (class: "SurfaceLight") -> (r.vSurfaceLight) -> SurfaceLight )
&	( (class: "LocalizationTable") -> (r.vLocalizationTable) -> LocalizationTable )
&	( (class: "Script") -> (r.vScript) -> Script )
&	( (class: "LocalScript") -> (r.vLocalScript) -> LocalScript )
&	( (class: "ModuleScript") -> (r.vModuleScript) -> ModuleScript )
&	( (class: "MarkerCurve") -> (r.vMarkerCurve) -> MarkerCurve )
&	( (class: "MaterialVariant") -> (r.vMaterialVariant) -> MaterialVariant )
&	( (class: "MemoryStoreService") -> (r.vMemoryStoreService) -> MemoryStoreService )
&	( (class: "Message") -> (r.vMessage) -> Message )
&	( (class: "Hint") -> (r.vHint) -> Hint )
&	( (class: "NoCollisionConstraint") -> (r.vNoCollisionConstraint) -> NoCollisionConstraint )
&	( (class: "OperationGraph") -> (r.vOperationGraph) -> OperationGraph )
&	( (class: "CornerWedgePart") -> (r.vCornerWedgePart) -> CornerWedgePart )
&	( (class: "Part") -> (r.vPart) -> Part )
&	( (class: "FlagStand") -> (r.vFlagStand) -> FlagStand )
&	( (class: "Seat") -> (r.vSeat) -> Seat )
&	( (class: "SkateboardPlatform") -> (r.vSkateboardPlatform) -> SkateboardPlatform )
&	( (class: "SpawnLocation") -> (r.vSpawnLocation) -> SpawnLocation )
&	( (class: "WedgePart") -> (r.vWedgePart) -> WedgePart )
&	( (class: "MeshPart") -> (r.vMeshPart) -> MeshPart )
&	( (class: "PartOperation") -> (r.vPartOperation) -> PartOperation )
&	( (class: "IntersectOperation") -> (r.vIntersectOperation) -> IntersectOperation )
&	( (class: "NegateOperation") -> (r.vNegateOperation) -> NegateOperation )
&	( (class: "UnionOperation") -> (r.vUnionOperation) -> UnionOperation )
&	( (class: "TrussPart") -> (r.vTrussPart) -> TrussPart )
&	( (class: "VehicleSeat") -> (r.vVehicleSeat) -> VehicleSeat )
&	( (class: "Model") -> (r.vModel) -> Model )
&	( (class: "Actor") -> (r.vActor) -> Actor )
&	( (class: "HopperBin") -> (r.vHopperBin) -> HopperBin )
&	( (class: "Tool") -> (r.vTool) -> Tool )
&	( (class: "Flag") -> (r.vFlag) -> Flag )
&	( (class: "WorldModel") -> (r.vWorldModel) -> WorldModel )
&	( (class: "PartOperationAsset") -> (r.vPartOperationAsset) -> PartOperationAsset )
&	( (class: "ParticleEmitter") -> (r.vParticleEmitter) -> ParticleEmitter )
&	( (class: "PathfindingLink") -> (r.vPathfindingLink) -> PathfindingLink )
&	( (class: "PathfindingModifier") -> (r.vPathfindingModifier) -> PathfindingModifier )
&	( (class: "Player") -> (r.vPlayer) -> Player )
&	( (class: "PluginAction") -> (r.vPluginAction) -> PluginAction )
&	( (class: "PluginCapabilities") -> (r.vPluginCapabilities) -> PluginCapabilities )
&	( (class: "NumberPose") -> (r.vNumberPose) -> NumberPose )
&	( (class: "Pose") -> (r.vPose) -> Pose )
&	( (class: "BloomEffect") -> (r.vBloomEffect) -> BloomEffect )
&	( (class: "BlurEffect") -> (r.vBlurEffect) -> BlurEffect )
&	( (class: "ColorCorrectionEffect") -> (r.vColorCorrectionEffect) -> ColorCorrectionEffect )
&	( (class: "DepthOfFieldEffect") -> (r.vDepthOfFieldEffect) -> DepthOfFieldEffect )
&	( (class: "SunRaysEffect") -> (r.vSunRaysEffect) -> SunRaysEffect )
&	( (class: "ProximityPrompt") -> (r.vProximityPrompt) -> ProximityPrompt )
&	( (class: "ProximityPromptService") -> (r.vProximityPromptService) -> ProximityPromptService )
&	( (class: "ReflectionMetadata") -> (r.vReflectionMetadata) -> ReflectionMetadata )
&	( (class: "ReflectionMetadataCallbacks") -> (r.vReflectionMetadataCallbacks) -> ReflectionMetadataCallbacks )
&	( (class: "ReflectionMetadataClasses") -> (r.vReflectionMetadataClasses) -> ReflectionMetadataClasses )
&	( (class: "ReflectionMetadataEnums") -> (r.vReflectionMetadataEnums) -> ReflectionMetadataEnums )
&	( (class: "ReflectionMetadataEvents") -> (r.vReflectionMetadataEvents) -> ReflectionMetadataEvents )
&	( (class: "ReflectionMetadataFunctions") -> (r.vReflectionMetadataFunctions) -> ReflectionMetadataFunctions )
&	( (class: "ReflectionMetadataClass") -> (r.vReflectionMetadataClass) -> ReflectionMetadataClass )
&	( (class: "ReflectionMetadataEnum") -> (r.vReflectionMetadataEnum) -> ReflectionMetadataEnum )
&	( (class: "ReflectionMetadataEnumItem") -> (r.vReflectionMetadataEnumItem) -> ReflectionMetadataEnumItem )
&	( (class: "ReflectionMetadataMember") -> (r.vReflectionMetadataMember) -> ReflectionMetadataMember )
&	( (class: "ReflectionMetadataProperties") -> (r.vReflectionMetadataProperties) -> ReflectionMetadataProperties )
&	( (class: "ReflectionMetadataYieldFunctions") -> (r.vReflectionMetadataYieldFunctions) -> ReflectionMetadataYieldFunctions )
&	( (class: "RemoteFunction") -> (r.vRemoteFunction) -> RemoteFunction )
&	( (class: "RenderingTest") -> (r.vRenderingTest) -> RenderingTest )
&	( (class: "RotationCurve") -> (r.vRotationCurve) -> RotationCurve )
&	( (class: "BuoyancySensor") -> (r.vBuoyancySensor) -> BuoyancySensor )
&	( (class: "ControllerPartSensor") -> (r.vControllerPartSensor) -> ControllerPartSensor )
&	( (class: "Sky") -> (r.vSky) -> Sky )
&	( (class: "Smoke") -> (r.vSmoke) -> Smoke )
&	( (class: "Sound") -> (r.vSound) -> Sound )
&	( (class: "ChorusSoundEffect") -> (r.vChorusSoundEffect) -> ChorusSoundEffect )
&	( (class: "CompressorSoundEffect") -> (r.vCompressorSoundEffect) -> CompressorSoundEffect )
&	( (class: "DistortionSoundEffect") -> (r.vDistortionSoundEffect) -> DistortionSoundEffect )
&	( (class: "EchoSoundEffect") -> (r.vEchoSoundEffect) -> EchoSoundEffect )
&	( (class: "EqualizerSoundEffect") -> (r.vEqualizerSoundEffect) -> EqualizerSoundEffect )
&	( (class: "FlangeSoundEffect") -> (r.vFlangeSoundEffect) -> FlangeSoundEffect )
&	( (class: "PitchShiftSoundEffect") -> (r.vPitchShiftSoundEffect) -> PitchShiftSoundEffect )
&	( (class: "ReverbSoundEffect") -> (r.vReverbSoundEffect) -> ReverbSoundEffect )
&	( (class: "TremoloSoundEffect") -> (r.vTremoloSoundEffect) -> TremoloSoundEffect )
&	( (class: "SoundGroup") -> (r.vSoundGroup) -> SoundGroup )
&	( (class: "Sparkles") -> (r.vSparkles) -> Sparkles )
&	( (class: "StandalonePluginScripts") -> (r.vStandalonePluginScripts) -> StandalonePluginScripts )
&	( (class: "StarterGear") -> (r.vStarterGear) -> StarterGear )
&	( (class: "StudioAttachment") -> (r.vStudioAttachment) -> StudioAttachment )
&	( (class: "StudioCallout") -> (r.vStudioCallout) -> StudioCallout )
&	( (class: "StyleRule") -> (r.vStyleRule) -> StyleRule )
&	( (class: "StyleSheet") -> (r.vStyleSheet) -> StyleSheet )
&	( (class: "StyleDerive") -> (r.vStyleDerive) -> StyleDerive )
&	( (class: "StyleLink") -> (r.vStyleLink) -> StyleLink )
&	( (class: "SurfaceAppearance") -> (r.vSurfaceAppearance) -> SurfaceAppearance )
&	( (class: "Team") -> (r.vTeam) -> Team )
&	( (class: "TeleportOptions") -> (r.vTeleportOptions) -> TeleportOptions )
&	( (class: "TerrainDetail") -> (r.vTerrainDetail) -> TerrainDetail )
&	( (class: "TerrainRegion") -> (r.vTerrainRegion) -> TerrainRegion )
&	( (class: "TestService") -> (r.vTestService) -> TestService )
&	( (class: "TextChannel") -> (r.vTextChannel) -> TextChannel )
&	( (class: "TextChatCommand") -> (r.vTextChatCommand) -> TextChatCommand )
&	( (class: "TextChatMessageProperties") -> (r.vTextChatMessageProperties) -> TextChatMessageProperties )
&	( (class: "TrackerStreamAnimation") -> (r.vTrackerStreamAnimation) -> TrackerStreamAnimation )
&	( (class: "Trail") -> (r.vTrail) -> Trail )
&	( (class: "Tween") -> (r.vTween) -> Tween )
&	( (class: "UIAspectRatioConstraint") -> (r.vUIAspectRatioConstraint) -> UIAspectRatioConstraint )
&	( (class: "UISizeConstraint") -> (r.vUISizeConstraint) -> UISizeConstraint )
&	( (class: "UITextSizeConstraint") -> (r.vUITextSizeConstraint) -> UITextSizeConstraint )
&	( (class: "UICorner") -> (r.vUICorner) -> UICorner )
&	( (class: "UIFlexItem") -> (r.vUIFlexItem) -> UIFlexItem )
&	( (class: "UIGradient") -> (r.vUIGradient) -> UIGradient )
&	( (class: "UIGridLayout") -> (r.vUIGridLayout) -> UIGridLayout )
&	( (class: "UIListLayout") -> (r.vUIListLayout) -> UIListLayout )
&	( (class: "UIPageLayout") -> (r.vUIPageLayout) -> UIPageLayout )
&	( (class: "UITableLayout") -> (r.vUITableLayout) -> UITableLayout )
&	( (class: "UIPadding") -> (r.vUIPadding) -> UIPadding )
&	( (class: "UIScale") -> (r.vUIScale) -> UIScale )
&	( (class: "UIStroke") -> (r.vUIStroke) -> UIStroke )
&	( (class: "UserNotification") -> (r.vUserNotification) -> UserNotification )
&	( (class: "UserNotificationPayload") -> (r.vUserNotificationPayload) -> UserNotificationPayload )
&	( (class: "UserNotificationPayloadAnalyticsData") -> (r.vUserNotificationPayloadAnalyticsData) -> UserNotificationPayloadAnalyticsData )
&	( (class: "UserNotificationPayloadJoinExperience") -> (r.vUserNotificationPayloadJoinExperience) -> UserNotificationPayloadJoinExperience )
&	( (class: "UserNotificationPayloadParameterValue") -> (r.vUserNotificationPayloadParameterValue) -> UserNotificationPayloadParameterValue )
&	( (class: "BinaryStringValue") -> (r.vBinaryStringValue) -> BinaryStringValue )
&	( (class: "BoolValue") -> (r.vBoolValue) -> BoolValue )
&	( (class: "BrickColorValue") -> (r.vBrickColorValue) -> BrickColorValue )
&	( (class: "CFrameValue") -> (r.vCFrameValue) -> CFrameValue )
&	( (class: "Color3Value") -> (r.vColor3Value) -> Color3Value )
&	( (class: "DoubleConstrainedValue") -> (r.vDoubleConstrainedValue) -> DoubleConstrainedValue )
&	( (class: "IntConstrainedValue") -> (r.vIntConstrainedValue) -> IntConstrainedValue )
&	( (class: "IntValue") -> (r.vIntValue) -> IntValue )
&	( (class: "NumberValue") -> (r.vNumberValue) -> NumberValue )
&	( (class: "ObjectValue") -> (r.vObjectValue) -> ObjectValue )
&	( (class: "RayValue") -> (r.vRayValue) -> RayValue )
&	( (class: "StringValue") -> (r.vStringValue) -> StringValue )
&	( (class: "Vector3Value") -> (r.vVector3Value) -> Vector3Value )
&	( (class: "Vector3Curve") -> (r.vVector3Curve) -> Vector3Curve )
&	( (class: "VirtualInputManager") -> (r.vVirtualInputManager) -> VirtualInputManager )
&	( (class: "WeldConstraint") -> (r.vWeldConstraint) -> WeldConstraint )
&	( (class: "Wire") -> (r.vWire) -> Wire )
